inference: don't add backdge when `applicable` inferred to return `Bool`

aviatesk opened the pull request:
~~Even if the return value of `applicable(f, args...)` is initially inferred as `Const(true)`, when a new method is added to `f` then it may cause new method match ambiguities, requiring the call to return return `false` at the next invocation. To handle such cases, it is necessary to always add a method table backedge in the inference of `applicable`.~~

Also just as a minor backedge reduction optimization, this commit avoids adding backedges when `applicable` is inferred to return `::Bool`.

vtjnash reviewed (changes requested):
These tests already appear to pass on master? The description doesn't make sense, as you cannot actually add this edge (even with force=true), as the mt wouldn't permit invalidating it if any method already exists (ambiguous or not)

aviatesk commented:
Yeah.. the newly added test cases seem to pass fine on `master` as well. It looks like the kind of invalidation I was concerned about is already covered by the method instance backedge. I’ll just keep the minor optimization to avoid adding a backedge when `rt === Bool`.​

aviatesk committed:
inference: don't add backedge when `applicable` inferred to return `Bool`

As a minor backedge reduction optimization, this commit avoids adding
backedges when `applicable` is inferred to return `::Bool`.
